### PR TITLE
Implement `--mode` and `--zarr-mode`; double-check Zarr checksums at end of backup

### LINKED ISSUE #284: Add --zarr-mode asset-checksum
https://github.com/dandi/dandisets/pull/282/files# implements `--mode timestamp` which relies on comparing stored for zarr latest timestamp across all keys for that prefix on S3. While doing a backup of a dandiset with lots of zarrs, that would then entail listing all zarrs contained in that dandiset (could be thousands) on S3 -- too expensive -- even if no asset for those zarrs has changed.  I think we need addition of another mode to use asset's time stamp (present on the server and also stored in .dandi/assets) so we do not sync if the asset for that zarr has not changed  since the last sync of that dandiset.

### PR DESCRIPTION
**Pull request by jwodder**
As detailed in https://github.com/dandisets/000108/issues/13.

Also closes #284.

### CODE REVIEW

File: tools/backups2datalad/__main__.py
Comment (yarikoptic) [inline tools/backups2datalad/__main__.py]:
```suggestion
    type=click.Choice(list(ZarrMode)),
```
?

File: tools/backups2datalad/adataset.py
Comment (yarikoptic) [inline tools/backups2datalad/adataset.py:465]:
```suggestion
    file: str
    key: str
```
since those just copy pasted or you have some intention for them?

File: tools/backups2datalad/adandi.py
Comment (yarikoptic) [inline tools/backups2datalad/adandi.py:415]: I am curious why this way and not overload like you did for `asset_type`?

File: tools/backups2datalad/adataset.py
Comment (yarikoptic) [inline tools/backups2datalad/adataset.py:230]:
```suggestion
        # rely on the fact that every data component of zarr folder is in annex
        # and we keep .dandi/ folder content directly in git
        async with aclosing(self.aiter_annexed_files()) as afiles:
```
isn't that right?

File: tools/backups2datalad/adataset.py
Comment (yarikoptic) [inline tools/backups2datalad/adataset.py:237]: I am yet to comprehend how ZCTree and its .add works , but just to make sure -- so the order of files spit out by aiter_annexed_files doesn't matter?

File: tools/backups2datalad/adandi.py
Comment (yarikoptic) [inline tools/backups2datalad/adandi.py:312]: what is this Config for? I don't see either of those two attributes used anywhere

File: tools/backups2datalad/adandi.py
Comment (yarikoptic) [inline tools/backups2datalad/adandi.py:257]: please add docstring describing why we need this class and what changes it got from the one in dandi.dandiapi.

**Review by yarikoptic — APPROVED** [review]
Did a pass over, left some comments but overall looks good to my tired eyes -- I guess we will discover in action!

File: tools/backups2datalad/adataset.py
Comment (jwodder) [reply tools/backups2datalad/adataset.py:465]: They're just there as a full record of the `git-annex find --json` output fields, in case they're ever needed, so that I don't have to experiment with git-annex again.

File: tools/backups2datalad/adandi.py
Comment (jwodder) [reply tools/backups2datalad/adandi.py:415]: I just copied the code from dandi-cli.  I don't recall why it was done that way, but it doesn't hurt.

File: tools/backups2datalad/adataset.py
Comment (jwodder) [reply tools/backups2datalad/adataset.py:230]: Yes.

File: tools/backups2datalad/adataset.py
Comment (jwodder) [reply tools/backups2datalad/adataset.py:237]: Correct.

File: tools/backups2datalad/adandi.py
Comment (jwodder) [reply tools/backups2datalad/adandi.py:312]: This configures pydantic so that the class can use the non-pydantic `RemoteDandiset` class as an attribute and so that instances can be contructed with `RemoteAsset(identifier=...)` instead of `RemoteAsset(alias_id=...)` (The latter setting's not used here; the `Config` was just copied from dandi-cli).

File: tools/backups2datalad/adandi.py
Comment (jwodder) [reply tools/backups2datalad/adandi.py:257]: Done: 8aa9912

File: tools/backups2datalad/adandi.py
Comment (yarikoptic) [inline tools/backups2datalad/adandi.py]:
```suggestion
    class Config:
        """Configuration for pydantic"""
        # To use the non-pydantic RemoteDandiset as an attribute and 
        # to be able tocreate instances with RemoteAsset(identifier=...) instead of
        # RemoteAsset(alias_id=...)
        allow_population_by_field_name = True
```
so I and others could discover its purpose later on

File: tools/backups2datalad/zarr.py
Comment (yarikoptic) [inline tools/backups2datalad/zarr.py:453]: we can end up in a use-case where checksum file would be annexed, content dropped, so symlink exist but not found if accessed, please add an explicit check if not a symlink, and if it is -- `raise RuntimeError(f"annexed file {self.repo / CHECKSUM_FILE} without content")` or alike to ease troubleshooting

Comment (jwodder) [reply tools/backups2datalad/zarr.py:453]: Are you sure that can happen?  Zarr backups are initialized with a `.dandi/.gitattributes` file to keep everything in `.dandi/` (which includes the checksum file) out of git-annex (See line 486).

File: tools/backups2datalad/__main__.py
Comment (yarikoptic) [inline tools/backups2datalad/__main__.py]: so what is the default if `default=None`?  similarly for `--zarr-mode`.  --help is not helpful on that:

```
  --mode [timestamp|verify|force]
                                  How to decide whether to back up a Dandiset.
                                  'timestamp' — only if timestamp of last
                                  backup is older than modified timestamp on
                                  server; 'force' — always; 'verify' — always,
                                  but error if there are any changes without a
                                  change to the timestamp
  --tags / --no-tags              Enable/disable creation of tags for releases
                                  [default: enabled]
  -w, --workers INTEGER           Number of workers to run in parallel
  --zarr-mode [timestamp|checksum|force]
                                  How to decide whether to back up a Zarr.
                                  'timestamp' — only if timestamp of last
                                  backup is older than some Zarr entry in
                                  S3;'checksum' — only if Zarr checksum is out
                                  of date or doesn't match expected value;
                                  'force' — always
```

Comment (jwodder) [reply tools/backups2datalad/__main__.py]: The true defaults are set in `config.py` (they're both `timestamp`), but (like all config options) they can be set in the config file, and then an explicit option given on the command line overrides the config file, but the latter is only implementable if the CLI default is `None`.

I'll make the defaults explicit in the `--help`.

File: tools/backups2datalad/zarr.py
Comment (yarikoptic) [reply tools/backups2datalad/zarr.py:453]: ok, good, let's hope that never happens.

File: tools/backups2datalad/__main__.py
Comment (yarikoptic) [reply tools/backups2datalad/__main__.py]: Thank you!

**Comment by yarikoptic** [comment]
hm, testing stalled or became too heavy?? (seems to run for over 59 min now in https://github.com/dandi/dandisets/actions/runs/3300016246 whenever before was done in under 20 minutes. any ideas on either it could be due to changes here or just general github slowneess?

I thought to wait until the end of CI runs, and then merge so I could run with `--mode force --zarr-mode checksum` to trigger fixups of the  "broken" zarrs over the weekend. (then for default mode of operation we would need https://github.com/dandi/dandisets/issues/284)

**Comment by jwodder** [comment]
@yarikoptic I suspect that https://github.com/dandi/dandi-archive/issues/1336 is making various parts of the tests fail, but I don't think the retrying of 500 errors should take that long.  I've restarted the tests.

**Comment by yarikoptic** [comment]
they had been running for over an hour, and Jake merged the fix I believe more recently.  Could you push an update to trigger a clean rerun of tests?

**Comment by jwodder** [comment]
@yarikoptic `HEAD` adjusted to trigger a CI rerun.

**Comment by yarikoptic** [comment]
Have you ran tests locally @jwodder ?  for me they stall after

```
❯ DANDI_API_KEY=SENSORED nox -e test -- --cov-report=xml -v
nox > Running session test
nox > Re-using existing virtual environment at .nox/test.
nox > python -m pip install -r backups2datalad.req.txt
nox > python -m pip install pytest pytest-cov 'datalad[tests]' zarr
nox > python -m pytest --cov-report=xml -v test_backups2datalad
================================================================================================ test session starts ================================================================================================
platform linux -- Python 3.10.7, pytest-7.1.3, pluggy-1.0.0 -- /home/yoh/proj/dandi/dandisets/tools/.nox/test/bin/python
cachedir: .pytest_cache
rootdir: /home/yoh/proj/dandi/dandisets/tools, configfile: setup.cfg
plugins: anyio-3.6.1, fail-slow-0.3.0, cov-3.0.0
collected 15 items                                                                                                                                                                                                  

test_backups2datalad/test_commands.py::test_backup_command PASSED                                                                                                                                             [  6%]
test_backups2datalad/test_commands.py::test_populate PASSED                                                                                                                                                   [ 13%]
```

if I run with logging set etc

```
DEBUG    backups2datalad:datasetter.py:95 Superdataset committed
FAILED                                                                                                                                                                                                        [ 13%]
------------------------------------------------------------------------------------------------- live log teardown -------------------------------------------------------------------------------------------------
DEBUG    httpx._client:_client.py:1706 HTTP Request: GET https://api-staging.dandiarchive.org/api/dandisets/205206/versions/?order= "HTTP/1.1 200 OK"
DEBUG    httpx._client:_client.py:1706 HTTP Request: DELETE https://api-staging.dandiarchive.org/api/dandisets/205206/ "HTTP/1.1 500 Internal Server Error"
WARNING  backups2datalad:aioutil.py:157 Retrying DELETE request to /dandisets/205206/ in 0.970448 seconds as it raised HTTPStatusError: Server error '500 Internal Server Error' for url 'https://api-staging.dandiarchive.org/api/dandisets/205206/'
For more information check: https://httpstatuses.com/500
DEBUG    httpx._client:_client.py:1706 HTTP Request: DELETE https://api-staging.dandiarchive.org/api/dandisets/205206/ "HTTP/1.1 500 Internal Server Error"
WARNING  backups2datalad:aioutil.py:157 Retrying DELETE request to /dandisets/205206/ in 2.060763 seconds as it raised HTTPStatusError: Server error '500 Internal Server Error' for url 'https://api-staging.dandiarchive.org/api/dandisets/205206/'
For more information check: https://httpstatuses.com/500
DEBUG    httpx._client:_client.py:1706 HTTP Request: DELETE https://api-staging.dandiarchive.org/api/dandisets/205206/ "HTTP/1.1 500 Internal Server Error"
WARNING  backups2datalad:aioutil.py:157 Retrying DELETE request to /dandisets/205206/ in 4.126623 seconds as it raised HTTPStatusError: Server error '500 Internal Server Error' for url 'https://api-staging.dandiarchive.org/api/dandisets/205206/'
For more information check: https://httpstatuses.com/500
DEBUG    httpx._client:_client.py:1706 HTTP Request: DELETE https://api-staging.dandiarchive.org/api/dandisets/205206/ "HTTP/1.1 500 Internal Server Error"
WARNING  backups2datalad:aioutil.py:157 Retrying DELETE request to /dandisets/205206/ in 8.245608 seconds as it raised HTTPStatusError: Server error '500 Internal Server Error' for url 'https://api-staging.dandiarchive.org/api/dandisets/205206/'
For more information check: https://httpstatuses.com/500
DEBUG    httpx._client:_client.py:1706 HTTP Request: DELETE https://api-staging.dandiarchive.org/api/dandisets/205206/ "HTTP/1.1 500 Internal Server Error"
WARNING  backups2datalad:aioutil.py:157 Retrying DELETE request to /dandisets/205206/ in 16.444056 seconds as it raised HTTPStatusError: Server error '500 Internal Server Error' for url 'https://api-staging.dandiarchive.org/api/dandisets/205206/'
For more information check: https://httpstatuses.com/500
DEBUG    httpx._client:_client.py:1706 HTTP Request: DELETE https://api-staging.dandiarchive.org/api/dandisets/205206/ "HTTP/1.1 500 Internal Server Error"
WARNING  backups2datalad:aioutil.py:157 Retrying DELETE request to /dandisets/205206/ in 33.539297 seconds as it raised HTTPStatusError: Server error '500 Internal Server Error' for url 'https://api-staging.dandiarchive.org/api/dandisets/205206/'
For more information check: https://httpstatuses.com/500
... and so on with increasing timeout...
```
and there it stands...  in /mnt/backup/dandi/heroku-logs/dandi-api-staging/20221024-1901.log
 on drogon I can see

<details>
<summary>this following oddly formatted traceback</summary> 

```shell
2022-10-24T23:34:46.635572+00:00 heroku[router]: at=info method=DELETE path="/api/dandisets/205206/" host=api-staging.dandiarchive.org request_id=f4d7864e-97e6-47b6-a0c0-762f594b7d5f fwd="76.24.253.1" dyno=web.1 connect=0ms service=31ms status=500 bytes=512 protocol=https
2022-10-24T23:34:46.634436+00:00 app[web.1]: [23:34:46] ERROR    Internal Server Error: /api/dandisets/205206/     log.py:241
2022-10-24T23:34:46.634454+00:00 app[web.1]: Traceback (most recent call last):
2022-10-24T23:34:46.634454+00:00 app[web.1]: File
2022-10-24T23:34:46.634454+00:00 app[web.1]: "/app/.heroku/python/lib/python3.10/site-packages
2022-10-24T23:34:46.634454+00:00 app[web.1]: /django/core/handlers/exception.py", line 55, in
2022-10-24T23:34:46.634455+00:00 app[web.1]: inner
2022-10-24T23:34:46.634455+00:00 app[web.1]: response = get_response(request)
2022-10-24T23:34:46.634455+00:00 app[web.1]: File
2022-10-24T23:34:46.634455+00:00 app[web.1]: "/app/.heroku/python/lib/python3.10/site-packages
2022-10-24T23:34:46.634455+00:00 app[web.1]: /django/core/handlers/base.py", line 197, in
2022-10-24T23:34:46.634456+00:00 app[web.1]: _get_response
2022-10-24T23:34:46.634456+00:00 app[web.1]: response = wrapped_callback(request,
2022-10-24T23:34:46.634456+00:00 app[web.1]: *callback_args, **callback_kwargs)
2022-10-24T23:34:46.634456+00:00 app[web.1]: File
2022-10-24T23:34:46.634457+00:00 app[web.1]: "/app/.heroku/python/lib/python3.10/site-packages
2022-10-24T23:34:46.634457+00:00 app[web.1]: /sentry_sdk/integrations/django/views.py", line
2022-10-24T23:34:46.634457+00:00 app[web.1]: 67, in sentry_wrapped_callback
2022-10-24T23:34:46.634457+00:00 app[web.1]: return callback(request, *args, **kwargs)
2022-10-24T23:34:46.634457+00:00 app[web.1]: File
2022-10-24T23:34:46.634457+00:00 app[web.1]: "/app/.heroku/python/lib/python3.10/site-packages
2022-10-24T23:34:46.634457+00:00 app[web.1]: /django/views/decorators/csrf.py", line 54, in
2022-10-24T23:34:46.634457+00:00 app[web.1]: wrapped_view
2022-10-24T23:34:46.634458+00:00 app[web.1]: return view_func(*args, **kwargs)
2022-10-24T23:34:46.634458+00:00 app[web.1]: File
2022-10-24T23:34:46.634458+00:00 app[web.1]: "/app/.heroku/python/lib/python3.10/site-packages
2022-10-24T23:34:46.634458+00:00 app[web.1]: /rest_framework/viewsets.py", line 125, in view
2022-10-24T23:34:46.634458+00:00 app[web.1]: return self.dispatch(request, *args,
2022-10-24T23:34:46.634459+00:00 app[web.1]: **kwargs)
2022-10-24T23:34:46.634459+00:00 app[web.1]: File
2022-10-24T23:34:46.634459+00:00 app[web.1]: "/app/.heroku/python/lib/python3.10/site-packages
2022-10-24T23:34:46.634459+00:00 app[web.1]: /rest_framework/views.py", line 509, in dispatch
2022-10-24T23:34:46.634459+00:00 app[web.1]: response = self.handle_exception(exc)
2022-10-24T23:34:46.634460+00:00 app[web.1]: File
2022-10-24T23:34:46.634460+00:00 app[web.1]: "/app/.heroku/python/lib/python3.10/site-packages
2022-10-24T23:34:46.634460+00:00 app[web.1]: /rest_framework/views.py", line 469, in
2022-10-24T23:34:46.634460+00:00 app[web.1]: handle_exception
2022-10-24T23:34:46.634460+00:00 app[web.1]: self.raise_uncaught_exception(exc)
2022-10-24T23:34:46.634460+00:00 app[web.1]: File
2022-10-24T23:34:46.634462+00:00 app[web.1]: "/app/.heroku/python/lib/python3.10/site-packages
2022-10-24T23:34:46.634462+00:00 app[web.1]: /rest_framework/views.py", line 480, in
2022-10-24T23:34:46.634462+00:00 app[web.1]: raise_uncaught_exception
2022-10-24T23:34:46.634462+00:00 app[web.1]: raise exc
2022-10-24T23:34:46.634463+00:00 app[web.1]: File
2022-10-24T23:34:46.634463+00:00 app[web.1]: "/app/.heroku/python/lib/python3.10/site-packages
2022-10-24T23:34:46.634463+00:00 app[web.1]: /rest_framework/views.py", line 506, in dispatch
2022-10-24T23:34:46.634463+00:00 app[web.1]: response = handler(request, *args, **kwargs)
2022-10-24T23:34:46.634463+00:00 app[web.1]: File "/app/dandiapi/api/views/dandiset.py",
2022-10-24T23:34:46.634463+00:00 app[web.1]: line 264, in destroy
2022-10-24T23:34:46.634463+00:00 app[web.1]: delete_dandiset(user=request.user,
2022-10-24T23:34:46.634464+00:00 app[web.1]: dandiset=self.get_object())
2022-10-24T23:34:46.634469+00:00 app[web.1]: File
2022-10-24T23:34:46.634469+00:00 app[web.1]: "/app/dandiapi/api/services/dandiset/__init__.py"
2022-10-24T23:34:46.634469+00:00 app[web.1]: , line 60, in delete_dandiset
2022-10-24T23:34:46.634470+00:00 app[web.1]: dandiset.delete()
2022-10-24T23:34:46.634470+00:00 app[web.1]: File
2022-10-24T23:34:46.634470+00:00 app[web.1]: "/app/.heroku/python/lib/python3.10/site-packages
2022-10-24T23:34:46.634470+00:00 app[web.1]: /django/db/models/base.py", line 1136, in delete
2022-10-24T23:34:46.634470+00:00 app[web.1]: collector.collect([self],
2022-10-24T23:34:46.634470+00:00 app[web.1]: keep_parents=keep_parents)
2022-10-24T23:34:46.634470+00:00 app[web.1]: File
2022-10-24T23:34:46.634470+00:00 app[web.1]: "/app/.heroku/python/lib/python3.10/site-packages
2022-10-24T23:34:46.634471+00:00 app[web.1]: /django/db/models/deletion.py", line 350, in
2022-10-24T23:34:46.634471+00:00 app[web.1]: collect
2022-10-24T23:34:46.634471+00:00 app[web.1]: raise ProtectedError(
2022-10-24T23:34:46.634471+00:00 app[web.1]: django.db.models.deletion.ProtectedError:
2022-10-24T23:34:46.634471+00:00 app[web.1]: ("Cannot delete some instances of model
2022-10-24T23:34:46.634471+00:00 app[web.1]: 'Dandiset' because they are referenced through
2022-10-24T23:34:46.634471+00:00 app[web.1]: protected foreign keys: 'ZarrArchive.dandiset'.",
2022-10-24T23:34:46.634471+00:00 app[web.1]: {<AssetPath: AssetPath object (322947)>,
2022-10-24T23:34:46.634471+00:00 app[web.1]: <AssetPath: AssetPath object (322943)>})
2022-10-24T23:34:46.635019+00:00 app[web.1]: 10.1.84.183 - - [24/Oct/2022:23:34:46 +0000] "DELETE /api/dandisets/205206/ HTTP/1.1" 500 145 "-" "backups2datalad (https://github.com/dandi/dandisets) httpx/0.22.0 CPython/3.10.7"
```
</details>

with critical piece

```
2022-10-24T23:34:46.634471+00:00 app[web.1]: ("Cannot delete some instances of model
2022-10-24T23:34:46.634471+00:00 app[web.1]: 'Dandiset' because they are referenced through
2022-10-24T23:34:46.634471+00:00 app[web.1]: protected foreign keys: 'ZarrArchive.dandiset'.",
```

so smells like either the fix was not deployed to the server or it is something else. 

@AlmightyYakob  and @jwodder please troubleshoot and do necessary fixes to the components needed so this PR tests pass.

**Comment by jwodder** [comment]
@yarikoptic I've filed an issue on the Archive repo with an MVCE: https://github.com/dandi/dandi-archive/issues/1344

**Comment by jwodder** [comment]
@yarikoptic The tests pass now.

**Comment by yarikoptic** [comment]
Woohoo, let's see how it goes ;-)

**Pull request merged**
